plugin/src/main/groovy/com/novoda/staticanalysis/internal/findbugs/FindbugsConfigurator.groovy

```diff
@@ -45,23 +50,71 @@ class FindbugsConfigurator extends CodeQualityConfigurator<FindBugs, FindBugsExt
             FindBugs task = project.tasks.create("findbugs${variant.name.capitalize()}", QuietFindbugsPlugin.Task)
+            List<File> androidSourceDirs = variant.sourceSets.collect { it.javaDirectories }.flatten()
             task.with {
-                group = "verification"
                 description = "Run FindBugs analysis for ${variant.name} classes"
-                source = variant.sourceSets.java.srcDirs.collect { it.path }.flatten()
-                classes = project.fileTree(variant.javaCompile.destinationDir)
+                source = androidSourceDirs
                 classpath = variant.javaCompile.classpath
-                dependsOn variant.javaCompile
+            }
+            sourceFilter.applyTo(task)
+            task.conventionMapping.map("classes", {
+                List<String> includes = createIncludesPatternsFrom(task.source, androidSourceDirs)
+                project.fileTree(variant.javaCompile.destinationDir).include(includes)
+            }.memoize());
+            task.dependsOn variant.javaCompile
+        }
+    }
+
+    @Override
+    protected void configureJavaProject() {
+        project.sourceSets.each { SourceSet sourceSet ->
+            String taskName = sourceSet.getTaskName(toolName, null)
+            FindBugs task = project.tasks.findByName(taskName)
+            if (task != null) {
+                sourceFilter.applyTo(task)
+                task.conventionMapping.map("classes", {
+                    List<File> sourceDirs = sourceSet.allJava.srcDirs.findAll { it.exists() }.toList()
+                    List<String> includes = createIncludesPatternsFrom(task.source, sourceDirs)
+                    createClassesTreeFrom(sourceSet).include(includes)
+                }.memoize());
             }
         }
     }
 
+    private static List<String> createIncludesPatternsFrom(FileCollection sourceFiles, List<File> sourceDirs) {
+        List<Path> includedSourceFiles = sourceFiles.matching { '**/*.java' }.files.collect { it.toPath() }
+        createIncludePatterns(includedSourceFiles, sourceDirs.collect { it.toPath() })
+    }
+
+    private static List<String> createIncludePatterns(List<Path> includedSourceFiles, List<Path> sourceDirs) {
+        createRelativePaths(includedSourceFiles, sourceDirs)
+                .collect { Path relativePath -> (relativePath as String) - '.java' + '*' }
+    }
+
+    private static List<Path> createRelativePaths(List<Path> includedSourceFiles, List<Path> sourceDirs) {
+        includedSourceFiles.collect { Path sourceFile ->
+            sourceDirs
+                    .findAll { Path sourceDir -> sourceFile.startsWith(sourceDir) }
+                    .collect { Path sourceDir -> sourceDir.relativize(sourceFile) }
+        }
+        .flatten()
+    }
+
+    /**
+     * The simple "classes = sourceSet.output" may lead to non-existing resources directory
+     * being passed to FindBugs Ant task, resulting in an error
+     * */
```

Comment:
Well, we are not changing the default behaviour here, instead we stack filters upon the default implementation: this method has been copied (and groovyfied) from the original [`FindbugsPlugin#L189-L191`](https://github.com/gradle/gradle/blob/master/subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/FindBugsPlugin.java#L189-L191). Should we reference the source?